nri-kafka/src/Kafka/Worker/Partition.hs

```diff
@@ -0,0 +1,503 @@
+import qualified Control.Exception.Safe as Exception
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString as ByteString
+import qualified Data.Sequence as Seq
+import qualified Data.Text.Encoding
+import qualified Data.Time.Clock as Clock
+import qualified Data.Time.Clock.POSIX as Clock.POSIX
+import qualified Data.UUID
+import qualified Data.UUID.V4
+import qualified GHC.Clock
+import qualified Kafka.Consumer as Consumer
+import qualified Kafka.Internal as Internal
+import qualified Kafka.Worker.Analytics as Analytics
+import qualified Kafka.Worker.Settings as Settings
+import qualified Kafka.Worker.Stopping as Stopping
+import qualified Log.Kafka
+import qualified Observability
+import qualified Platform
+import qualified Prelude
+
+data WorkerError e
+  = WorkerCallbackFailed e
+  | WorkerCallbackThrew Exception.SomeException
+  | MsgDecodingFailed Text
+  | SeekFailed Consumer.KafkaError
+  deriving (Show)
+
+data State = State
+  { analytics :: Analytics.Analytics,
+    stopping :: Stopping.Stopping,
+    partition :: Partition
+  }
+
+type ConsumerRecord = Consumer.ConsumerRecord (Maybe ByteString.ByteString) (Maybe ByteString.ByteString)
+
+newtype ProcessAttemptsCount = ProcessAttemptsCount Int
+  deriving (Num)
+
+newtype Partition = Partition (TVar.TVar Backlog)
+
+-- | ProcessResult is the expected response of the MessageCallback.
+-- Return Success if processing succeeded and the offset was correct.
+-- Return ExpectedOffset with the expected offset if the offset of the message
+-- processed was wrong (this only makes sense when Kafka isn't managing the offset)
+-- The MessageCallback will throw if proccessing throws.
+data ProcessResult
+  = Success
+  | ExpectedOffset Int
+
+-- for each partition, we keep a local backlog of messages we've read from Kafka
+-- that are not yet processed
+data Backlog
+  = AwaitingSeekTo Int
+  | -- | Use a `Sequence` type to store records. It's a bit like a list, but unlike
+    -- lists it has O(1) appends to both ends, and an O(1) length function as well.
+    -- We use this a lot considering we're appending new messages on one end, and
+    -- processes messages from the other.
+    Assigned (Seq.Seq (ProcessAttemptsCount, ConsumerRecord))
+  | -- we set the Backlog to this to tell the partition's thread to stop
+    Stopping
+
+-- using these types as a proxy for named parameters
+-- this is a function that runs with the partition once it's started
+newtype OnStartup = OnStartup (Partition -> Prelude.IO ())
+
+-- using this types as a proxy for named parameters
+-- this is a function that runs with the partition when it's done to cleanup
+newtype OnCleanup = OnCleanup (Prelude.IO ())
+
+-- | MessageCallback is used by your worker to process messages peeled off the queue.
+data MessageCallback where
+  MessageCallback ::
+    (Show e, Aeson.ToJSON msg, Aeson.FromJSON msg) =>
+    (Consumer.ConsumerRecord () () -> msg -> Task e ProcessResult) ->
```

Comment:
:art: Might be nice to pass out a custom thing instead of `Consumer.ConsumerRecord () ()` here, so we can control documentation and don't require library users to use these kafka wrapper types.